Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp

```diff
@@ -0,0 +1,292 @@
+	bInitializedCustomSpawnParameters = true;
+}
+
+bool ABenchmarkGymGameMode::ShouldUseCustomSpawning()
+{
+	FString WorkerValue;
+	USpatialWorkerFlags::GetWorkerFlag(TEXT("override_spawning"), WorkerValue);
+	return (WorkerValue.Equals(TEXT("true"), ESearchCase::IgnoreCase) || FParse::Param(FCommandLine::Get(), TEXT("OverrideSpawning")));
+}
+
+void ABenchmarkGymGameMode::ParsePassedValues()
+{
+	if (FParse::Param(FCommandLine::Get(), TEXT("OverrideSpawning")))
+	{
+		UE_LOG(LogBenchmarkGym, Log, TEXT("Found OverrideSpawning in command line args, worker flags for custom spawning will be ignored."));
+		FParse::Value(FCommandLine::Get(), TEXT("TotalPlayers="), TotalPlayers);
+
+		// Set default value of PlayerDensity equal to TotalPlayers. Will be overwritten if PlayerDensity option is specified.
+		PlayerDensity = TotalPlayers;
+
+		FParse::Value(FCommandLine::Get(), TEXT("PlayerDensity="), PlayerDensity);
+
+		FParse::Value(FCommandLine::Get(), TEXT("TotalNPCs="), TotalNPCs);
+	}
+	else
+	{
+		UE_LOG(LogBenchmarkGym, Log, TEXT("Using worker flags to load custom spawning parameters."));
+		FString TotalPlayersString, PlayerDensityString, TotalNPCsString;
+		if (USpatialWorkerFlags::GetWorkerFlag(TEXT("total_players"), TotalPlayersString))
+		{
+			TotalPlayers = FCString::Atoi(*TotalPlayersString);
+		}
+
+		// Set default value of PlayerDensity equal to TotalPlayers. Will be overwritten if PlayerDensity option is specified.
+		PlayerDensity = TotalPlayers;
+
+		if (USpatialWorkerFlags::GetWorkerFlag(TEXT("player_density"), PlayerDensityString))
+		{
+			PlayerDensity = FCString::Atoi(*PlayerDensityString);
+		}
+
+		if (USpatialWorkerFlags::GetWorkerFlag(TEXT("total_npcs"), TotalNPCsString))
+		{
+			TotalNPCs = FCString::Atoi(*TotalNPCsString);
+		}
+	}
+
+	NumPlayerClusters = FMath::CeilToInt(TotalPlayers / (float)PlayerDensity);
+}
+
+void ABenchmarkGymGameMode::ClearExistingSpawnPoints()
+{
+	UGameplayStatics::GetAllActorsOfClass(GetWorld(), APlayerStart::StaticClass(), SpawnPoints);
+	for (AActor* SpawnPoint : SpawnPoints)
+	{
+		SpawnPoint->Destroy();
+	}
+}
+
+void ABenchmarkGymGameMode::GenerateGridSettings(int DistBetweenPoints, int NumPoints, int& NumRows, int& NumCols, int& MinRelativeX, int& MinRelativeY)
+{
+	if (NumPoints <= 0)
+	{
+		UE_LOG(LogBenchmarkGym, Warning, TEXT("Generating grid settings with non-postive number of points (%d)"), NumPoints);
+		NumRows = 0;
+		NumCols = 0;
+		MinRelativeX = 0;
+		MinRelativeY = 0;
+		return;
+	}
+
+	NumRows = FMath::RoundToInt(FMath::Sqrt(NumPoints));
+	NumCols = FMath::CeilToInt(NumPoints / (float)NumRows);
+	const int GridWidth = (NumCols - 1) * DistBetweenPoints;
+	const int GridHeight = (NumRows - 1) * DistBetweenPoints;
+	MinRelativeX = FMath::RoundToInt(-GridWidth / 2.0);
+	MinRelativeY = FMath::RoundToInt(-GridHeight / 2.0);
+}
+
+void ABenchmarkGymGameMode::GenerateSpawnPointClusters(int NumClusters)
+{
+	const int DistBetweenClusterCenters = 40000; // 400 meters, in Unreal units.
```

Comment:
Could make this static? I discussed this with Sami the other day and it was kind of undecided what we wanted to do here soo.... do whatever you feel is best. If you do go with static there are a few other places where you could add it.